File: app/src/main/java/com/example/muzik/response_model/Song.java
package com.example.muzik.response_model;

import android.net.Uri;

import androidx.annotation.Nullable;

import com.example.muzik.music_service.model.Album;
import com.example.muzik.music_service.model.Artist;

public class Song extends com.example.muzik.music_service.model.Song implements ResponseModel {
    private int songID;
    private Integer size;
    private String artistName;
    private Long artistID;
    private String imageURL;
    private Integer albumID;
    private String songURL;
    private int views;

    public Song(boolean isNewSample) {
        super(null, null, null);
        if (isNewSample) songID = -1;
    }

    public String getSongURL() {
        return songURL;
    }

    public int getSongID() {
        return songID;
    }

    public void setSongID(int songID) {
        this.songID = songID;
    }

    public Integer getSize() {
        return size;
    }

    public void setSize(Integer size) {
        this.size = size;
    }

    public Long getArtistID() {
        return artistID;
    }

    public void setArtistID(Long artistID) {
        this.artistID = artistID;
    }

    public String getImageURL() {
        return imageURL;
    }

    public void setImageURL(String imageURL) {
        this.imageURL = imageURL;
    }

    public Integer getAlbumID() {
        return albumID;
    }

    public void setAlbumID(Integer albumID) {
        this.albumID = albumID;
    }

    public String getArtistName() {
        return artistName;
    }


    public void setArtistName(String artistName) {
        this.artistName = artistName;
    }

    public void setSongURL(String songURL) {
        this.songURL = songURL;
    }

    public int getViews() {
        return views;
    }

    public void setViews(int views) {
        this.views = views;
    }

    @Nullable
    @Override
    public Artist getArtist() {
        return null;
    }

    @Nullable
    @Override
    public Uri getImg() {
        return null;
    }

    @Nullable
    @Override
    public Album getAlbum() {
        return null;
    }
}

File: app/src/main/java/com/example/muzik/data_model/standard_model/Song.kt
package com.example.muzik.data_model.standard_model

import android.net.Uri
import com.example.muzik.data_model.retrofit_model.response.SongResponse
import com.example.muzik.music_service.LocalMusicRepository

class Song(
    val songID: String? = null,
    val name: String? = null,
    var artistName: String? = null,
    var songURI: Uri? = null,
    val views: Long? = -1,
    val imageURI: Uri? = null,
    val artistID: Long? = -1,
    var duration: Int = -1,
    val size: Int? = -1,
    val albumID: Long? = -1
) : Model {

    companion object {
        fun buildOnline(songResponse: SongResponse): Song {
            var songURI: Uri? = null
            songResponse.songURL?.let {
                //songURI = Uri.parse(it.replace("http:/", "https://"))
                songURI = Uri.parse(it)
            }
            return Song(
                songID = songResponse.songID,
                name = songResponse.name,
                artistName = songResponse.artistName,
                songURI = songURI,
                views = songResponse.views,
                imageURI = Uri.parse(songResponse.imageURL),
                artistID = songResponse.artistID,
                duration = songResponse.duration
            )
        }

        fun buildLocal(
            songId: Long,
            uri: Uri,
            displayName: String,
            duration: Int,
            size: Int,
            albumID: Long,
            artistID: Long
        ): Song {
            val artist = LocalMusicRepository.getArtist(artistID)
                ?: Artist.buildLocal(-1, "Unknown")
            return Song(
                songID = songId.toString(),
                name = displayName,
                songURI = uri,
                artistID = artistID,
                duration = duration,
                size = size,
                albumID = albumID,
                artistName = artist.name
            )
        }
    }

    fun getAlbum(): Album {
        return albumID?.let { LocalMusicRepository.getAlbum(it) } ?: Album.buildLocal(
            -1,
            "Unknown",
            Uri.parse("default"),
            -1
        )
    }

    fun requireSongID(): String {
        checkNotNull(songID) {
            ("Song $this must have non-null songID")
        }
        return songID
    }

    fun requireArtistID(): Long {
        checkNotNull(artistID) {
            ("Song $this must have non-null artist ID")
        }
        return artistID
    }
}